kafka/consumer.go

```diff
@@ -182,20 +197,18 @@ func (c *Consumer) Run(ctx context.Context) error {
 // fetch polls the Kafka broker for new records up to cfg.MaxPollRecords.
 // Any errors returned by fetch should be considered fatal.
 func (c *Consumer) fetch(ctx context.Context) error {
-	// We need to grab the lock before pulling records to guarantee no events loss on shutdown.
-	c.mu.RLock()
-	defer c.mu.RUnlock()
-
 	fetches := c.client.PollRecords(ctx, c.cfg.MaxPollRecords)
 	defer c.client.AllowRebalance()
 
-	if fetches.IsClientClosed() {
-		return fmt.Errorf("client is closed: %w", context.Canceled)
-	}
-	if errors.Is(fetches.Err0(), context.Canceled) ||
+	if fetches.IsClientClosed() ||
+		errors.Is(fetches.Err0(), context.Canceled) ||
 		errors.Is(fetches.Err0(), context.DeadlineExceeded) {
-		return fmt.Errorf("context done: %w", fetches.Err0())
+		return context.Canceled
 	}
+	// Acquire the lock after checking if the the context is cancelled and/or
+	// the client is closed.
```

Comment:
**issue**: I don't think we can do this. What happens if the context is canceled or the client is closed before we acquire the lock but after the records were polled ? This is also making the previous if not safe for concurrent usage as the client could be closed after the `IsClientClosed` is called.

Reply:
> I don't think we can do this. What happens if the context is canceled or the client is closed before we acquire the lock but after the records were polled ? 

If the context is cancelled before the lock is acquired, the call on line 217 will fail for `AtMostOnceDelivery`, and the function will return then. No records will be lost since the commit for the fetched offsets will have failed.

If the context is cancelled between line 228 and 240, the records will be lost, since `AtMostOnceDelivery` has that caveat. This is no different with the lock acquired at the top of the function, before `PollRecords` is called, or after.

The locking doesn't prevent the context from being cancelled at any point in the execution flow.

> This is also making the previous if not safe for concurrent usage as the client could be closed after the IsClientClosed is called.

Why would closing the `kgo.Client` make calls to it not safe? The client can be called at any time, and the `kgo.Client` is safe for concurrent use. The client can be closed at any time while other methods are called, and is within the intended use scope.

Reply:
TIL! Thanks for the explanation! bow

My only other concern would be that the consumer is not guaranteed to return nil on shutdown. For example, in one of the case you mentioned:

    If the context is cancelled before the lock is acquired, the call on line 217 will fail for AtMostOnceDelivery, and the function will return then.

the consumer would return ErrCommitFailed. WDYT ?

Reply:
I think that's a legitimate error, since in that case, we may want to re-create the consumer, rather than an entire process.